Sources/AsyncHTTPClient/AsyncAwait/Transaction+StateMachine.swift

```diff
@@ -0,0 +1,603 @@
+
+            case .executing(let executor, let requestState, .waitingForRemote(let streamID, let continuation)):
+                var buffer = buffer
+                let first = buffer.removeFirst()
+                self.state = .executing(executor, requestState, .buffering(streamID, buffer, next: .askExecutorForMore))
+                return .succeedContinuation(continuation, first)
+
+            case .finished:
+                // the request failed or was cancelled before, we can ignore further data
+                return .none
+
+            case .executing(_, _, .finished):
+                preconditionFailure("Received response end. Must not receive further body parts after that. Invalid state: \(self.state)")
+            }
+        }
+
+        enum ResponseBodyDeinitedAction {
+            case cancel(HTTPRequestExecutor)
+            case none
+        }
+
+        mutating func responseBodyDeinited() -> ResponseBodyDeinitedAction {
+            switch self.state {
+            case .initialized,
+                 .queued,
+                 .executing(_, _, .initialized):
+                preconditionFailure("Invalid state: \(self.state)")
+
+            case .executing(_, _, .waitingForStream(_, next: .endOfFile)):
+                self.state = .finished(error: nil, nil)
+                return .none
+
+            case .executing(let context, _, .waitingForStream(_, next: .askExecutorForMore)):
+                self.state = .finished(error: nil, nil)
+                return .cancel(context.executor)
+
+            case .executing(_, _, .waitingForStream(_, next: .error(let error))):
+                self.state = .finished(error: error, nil)
+                return .none
+
+            case .finished:
+                // body was released after the response was consumed
+                return .none
+
+            case .executing(_, _, .buffering),
+                 .executing(_, _, .waitingForRemote),
+                 .executing(_, _, .finished):
+                // user is consuming the stream with an iterator
+                return .none
+            }
+        }
+
+        mutating func cancelResponseStream(streamID: HTTPClientResponse.Body.IteratorStream.ID) -> FailAction {
+            switch self.state {
+            case .initialized, .queued, .executing(_, _, .initialized):
+                preconditionFailure("Got notice about a deinited iterator, before we even received a response. Invalid state: \(self.state)")
+
+            case .executing(_, _, .buffering(let registeredStreamID, _, next: _)),
+                 .executing(_, _, .waitingForRemote(let registeredStreamID, _)):
+                guard registeredStreamID == streamID else {
+                    // we got informed about an iterator that we don't actually serve!
+                    return .none
+                }
+                return self.fail(HTTPClientError.cancelled)
+
+            case .executing(_, _, .waitingForStream),
```

Comment:
Uh, is `.waitingForStream` equivalent to `.finished` here? Why isn't this an error?

Reply:
Renamed `.waitingForStream` to `.waitingForResponseIterator`. To make its meaning clearer.